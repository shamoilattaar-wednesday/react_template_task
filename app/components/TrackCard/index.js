/**
 *
 * RepoCard
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { Card } from '@mui/material';
import T from '@components/T';
import If from '@components/If';
import isEmpty from 'lodash/isEmpty';

const CustomCard = styled(Card)`
  && {
    margin: 1rem 0;
    padding: 1rem;
  }
`;

export function TrackCard({ collectionName, artistName, collectionPrice, artworkUrl100 }) {
  return (
    <CustomCard data-testid="repo-card">
      
      <If condition={!isEmpty(collectionName)} otherwise={<T data-testid="name-unavailable" id="repo_name_unavailable" />}>
        <div ><img src={artworkUrl100} alt={collectionName} />
        <T data-testid="name" id="track_name" values={{ trackName: collectionName }} />
        </div>
       
      </If>
      <If
        condition={!isEmpty(artistName)}
        otherwise={<T data-testid="fullName-unavailable" id="repo_full_name_unavailable" />}
      >
        <T data-testid="fullName" id="artist_name" values={{ artist_name: artistName }} />
      </If>
      <If condition={collectionPrice} otherwise={<T data-testid="stargazers-unavaiable" id="repo_stars_unavailable" />}>
        <T data-testid="stargazers" id="total_cost" values={{ price: collectionPrice }} />
      </If>
    </CustomCard>
  );
}

TrackCard.propTypes = {
  collectionName: PropTypes.string,
  artistName: PropTypes.string,
  collectionPrice: PropTypes.number
};

export default TrackCard;
