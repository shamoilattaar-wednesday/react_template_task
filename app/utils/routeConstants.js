export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  tracks:{
    route: '/tracks',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  track:{
    route: '/tracks/:trackId',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  }
  
};
