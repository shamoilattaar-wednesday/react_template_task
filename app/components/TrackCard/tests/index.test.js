import React from 'react';
import { renderWithIntl, renderProvider, timeout } from '@utils/testUtils';
import TrackCard from '../index';
import { translate } from '@app/utils/index';
import { fireEvent } from '@testing-library/dom';

describe('<TrackCard />', () => {
  let submitSpy;
  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const track = {};
    const { baseElement } = renderWithIntl(<TrackCard track={track} />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should contain 1 TrackCard component', () => {
    const track = {};
    const { getAllByTestId } = renderWithIntl(<TrackCard track={track} />);
    expect(getAllByTestId('track-card').length).toBe(1);
    expect(getAllByTestId('track-details').length).toBe(1);
  });
  it('should render the track details inside the card', () => {
    const trackName = 'mac';
    const artistName = 'wednesday-solutions';
    const collectionPrice = 200;
    const track = { trackName: 'mac' };
    const { getByTestId } = renderWithIntl(
      <TrackCard collectionName={trackName} artistName={artistName} collectionPrice={collectionPrice} track={track} />
    );
    expect(getByTestId('name')).toHaveTextContent(trackName);
    expect(getByTestId('artist-name')).toHaveTextContent(artistName);
  });
  it('should render the repository unavailable messages in case any props are unavailable or have falsy values', () => {
    const repoUnavailable = translate('repo_name_unavailable');
    const fullNameUnavailable = translate('repo_full_name_unavailable');
    const stargazersUnavailable = translate('repo_stars_unavailable');
    const track = {};
    const { getByTestId } = renderWithIntl(<TrackCard track={track} />);
    expect(getByTestId('name-unavailable')).toHaveTextContent(repoUnavailable);
    expect(getByTestId('fullName-unavailable')).toHaveTextContent(fullNameUnavailable);
    expect(getByTestId('stargazers-unavaiable')).toHaveTextContent(stargazersUnavailable);
  });
  it('should render the track details from the click on card', async () => {
    const trackName = 'mac';
    const artistName = 'wednesday-solutions';
    const collectionPrice = 200;
    const track = {
      artworkUrl100: 'https://via.placeholder.com/100',
      collectionName: 'Test Collection',
      artistName: 'Test Artist',
      description: 'Test Description',
      collectionPrice: 9.99,
      trackName: 'mac',
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a'
    };
    const { getByTestId } = renderProvider(
      <TrackCard
        viewTracks={submitSpy}
        collectionName={trackName}
        artistName={artistName}
        collectionPrice={collectionPrice}
        track={track}
      />
    );
    fireEvent.click(getByTestId('track-details'));

    await timeout(500);
    expect(submitSpy).toBeCalledWith(track);
  });
  it('should render the track details from the click on play audio', async () => {
    const trackName = 'mac';
    const artistName = 'wednesday-solutions';
    const collectionPrice = 200;
    const track = {
      artworkUrl100: 'https://via.placeholder.com/100',
      collectionName: 'Test Collection',
      artistName: 'Test Artist',
      description: 'Test Description',
      collectionPrice: 9.99,
      trackName: 'mac',
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a'
    };
    const { getByTestId } = renderProvider(
      <TrackCard
        handlePlay={submitSpy}
        collectionName={trackName}
        artistName={artistName}
        collectionPrice={collectionPrice}
        track={track}
      />
    );
    fireEvent.click(getByTestId('play-audio'));

    await timeout(500);
    expect(submitSpy).toBeCalledWith(track);
  });
  it('should render the track details from the click on pause audio', async () => {
    const trackName = 'mac';
    const artistName = 'wednesday-solutions';
    const collectionPrice = 200;
    const isPlaying = true;
    const track = {
      artworkUrl100: 'https://via.placeholder.com/100',
      collectionName: 'Test Collection',
      artistName: 'Test Artist',
      description: 'Test Description',
      collectionPrice: 9.99,
      trackName: 'mac',
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/4a/68/06/4a6806aa-be22-3eff-8153-d9b1ab9dab8b/mzaf_12411707066990307220.plus.aac.p.m4a'
    };
    const { getByTestId } = renderProvider(
      <TrackCard
        handlePlay={submitSpy}
        collectionName={trackName}
        artistName={artistName}
        collectionPrice={collectionPrice}
        track={track}
        isPlaying={isPlaying}
      />
    );
    fireEvent.click(getByTestId('pause-audio'));

    await timeout(500);
    expect(submitSpy).toBeCalledWith(track);
  });
});
