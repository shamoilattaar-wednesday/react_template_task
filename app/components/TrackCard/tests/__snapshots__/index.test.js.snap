// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e12nrugs0 css-tcnzq6-MuiPaper-root-MuiCard-root"
      data-testid="track-card"
    >
      <div
        data-testid="track-details"
      >
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="name-unavailable"
        >
          Repository name is unavailable
        </p>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="fullName-unavailable"
        >
          Repository full name unavaiable
        </p>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="stargazers-unavaiable"
        >
          Repository stars are unavaiable
        </p>
      </div>
      <button
        data-testid="play-audio"
      >
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="t"
        >
          Play Audio
        </p>
      </button>
    </div>
  </div>
</body>
`;
